README.md

```diff
@@ -1,16 +1,1831 @@
+send.receipt.sig --> send.receipt.ran
+send.receipt.sig --> send.receipt.out
+send.receipt.sig --> send.receipt.fx
+send.receipt.sig --> send.receipt.meta
+send.receipt.sig --> send.receipt.iss
+send.receipt.sig --> send.receipt.prf
+
+
+send.receipt.fx --> send.fx.fork
+send.receipt.fx --> send.fx.join
+send.fx.fork --> update.task
+send.fx.fork --> read.task
+read.cause --> send
+update.cause --> send
+send.receipt.ran --> send
+```
+
+### 2.2.1 Task
+
+A [Task] is like a deferred function application: a request to perform some action on a resource with specific input.
+
+### 2.2.2 Authorization
+
+An [Authorization] is a cryptographically signed proof permitting execution of referenced tasks. It allows invoker to authorize a group of tasks using one cryptographic signature.
+
+### 2.2.3 Invocation
+
+An [Invocation] is an invoker authorized instruction to the [Executor] to run the [Task].
+
+### 2.2.4 Result
+
+A [Result] is the output of a [Task].
+
+### 2.2.5 Receipt
+
+A [Receipt] is a cryptographically signed description of the [Invocation] output.
+
+### 2.2.6 Effect
+
+An [Effect] are the instruction to the [Executor] to run set of [Task]s concurrently.
+
+## 2.3 IPLD Schema
+
+```ipldsch
+type Task struct {
+  with    URI
+  do      Ability
+  input   {String: any}
+
+  nnc     string
+}
+
+type URI string
+type Ability string
+
+type Authorization struct {
+  # Authorization is denoted by the set of links been authorized
+  scope   [&Any]
+  # Scope signed by the invoker
+  s       VarSig
+}
+
+type Invocation struct {
+  v       SemVer
+
+  run     &Task
+  # Receipt of the invocation that caused this invocation
+  cause   optional &Invocation
+  # Task authorization.
+  auth    &Authorization
+
+  meta    {String : any}
+
+  prf     [&UCAN]
+}
+type SemVer string
+
+type Receipt struct {
+  # Invocation this is a receipt for
+  ran     &Invocation
+
+  # Output of the invocation
+  out     Result
+
+  # Effects to be performed
+  fx      Effect
+
+  # All the other metadata
+  meta    {String: any}
+
+  # Principal that issued this receipt. If omitted issuer is
+  # inferred from the invocation task audience.
+  iss     optional Principal
+
+  # When issuer is different from executor this MUST hold a UCAN
+  # delegation chain from executor to the issuer. Should be omitted when
+  # executor is an issuer.
+  prf     [&UCAN]
+
+  # Signature from the "iss".
+  s       Varsig
+}
+
+type Result union {
+  | any    "ok"    # Success
+  | any    "error" # Error
+} representation kinded
+
+# Represents a request to invoke enclosed set of tasks concurrently
+type Effect struct {
+  # Primary set of tasks to be invoked
+  fork      [&Invocation]
+  
+  # Additional task to be invoked with added semantics
+  # of representing a workflow execution continuation.
+  join       optional &Promise
+}
+
+# Promise is an Invocation with optional 'auth' field which if omitted
+# is implicitly an 'auth' of the Invocation that contains Await.
```

Comment:
I'm feeling like maybe I'm a bit slow on the uptake today 😅 

Aside from the optional auth field, this not identical to an `Invocation`, yes? Why not use the batch signature capability to sign such an inner Invocation? I'm also unclear what the user case is for this versus a Task.

Reply:
FWIW in IPVM, our need for the `await` is to point at a `Task` or `Invocation`.

Reply:
Ah, maybe you're going for this being equivalent to an `Invocation`, which makes sense and works for me — but I'm still curious about why the `optional auth`

Reply:
It is basically an `Invocation` yes only difference is that it's `auth` is optional. It allows you to await between invocations that are part of same `auth` group.

Without optional `auth` we'd have cycle where we have an await which refers to the Invocation which needs to refer to the `auth`.

In the past we had a tagged union of `Invocation` and `Task` so we could use later to imply same `auth` group as the Task awaiting. However after our chat last week we've decided we could just make `auth` optional in the `Invocation` instead and avoid tagging.

In the end I decided to keep `auth` mandatory on the `Invocation` and define `Promise` which has `auth` optional, so that it's only limited to `await`.

Reply:
In other words the reasons I have defined a `Promise` are:

1. So that whole pipeline could be authorized with a single `auth` (This is requirement from you, I'm ok with requiring that awaited tasks been signed separately from task awaiting, that would make bunch of things simpler)
2. So that `Invocation` requires `auth`, otherwise there may be `Invocations` without `auth` which would be invalid.

Reply:
> So that whole pipeline could be authorized with a single auth (This is requirement from you, I'm ok with requiring that awaited tasks been signed separately from task awaiting, that would make bunch of things simpler)

I think we're both trying to move quickly and are missing points from each other. If I have the ability to sign over multiple things via Varsig, I don't need things like `Batch`es, and can move an additional semantics to other layers.

However, I _do_ need the ability to create effects that are at least isomorphic to `Task`s — which is to say not wrapped in an `Invocation` for semantic reasons

Reply:
Like my comment higher up in this PR, I now remember that this is meant as being equivalent to `data Promise = Task | Invocation`, so I'm fine with it if it makes implementation easier. We can always revisit if it's unclear to others. I'll perhaps add a `suggestion` with some text to this effect in the schema comment.

Reply:
I'm  bit confused what you're getting at. We could remove `Promise` entirely and just make `auth` optional on the `Invocation`. I believe that was the plan too. Only reason I decided to define `Promise` with optional `auth` instead of making `auth` optional on `Invocation` is so that in some places where `auth` is required I would use `Invocation` (e.g. receipts) and in other cases where `auth` isn't required I could use `Promise` does that make sense ?

Reply:
Regarding https://github.com/ucan-wg/invocation/pull/10/files#r1099617974 comment, this is precise example illustrating what I was trying to say https://observablehq.com/@gozala/ucan-invocation-spec#batchedPipelineFlow

Here's actual js code

```js
  const updateDNS = crud.update("dns:example.com?TYPE=TXT", {
    value: "hello world"
  });

  const notifyBob = msg.send("mailto://alice@example.com", {
    to: "bob@example.com",
    subject: "DNSLink for example.com",
    body: updateDNS.await().ok
  });

  const notifyCarol = msg.send("mailto://alice@example.com", {
    to: "carol@example.com",
    subject: "Hey Carol, DNSLink was updated!",
    body: updateDNS.await().ok
  });

  const logDone = crud.update("https://example.com/report", {
    payload: {
      from: "mailto://alice@exmaple.com",
      to: ["bob@exmaple.com", "carol@example.com"],
      event: "email-notification"
    },
    _: [notifyBob.await().ok, notifyCarol.await().ok]
  });
```

If all were to be authorized with the same auth, only `logDone` could have pointer to the `auth` all the other invocations would have to be promises